core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,355 @@
+      return maxContentLength;
+    }
+
+    public long maxTotalBytes() {
+      return maxTotalBytes;
+    }
+
+    public Cache<String, CacheEntry> cache() {
+      return cache;
+    }
+
+    public CacheStats stats() {
+      return cache.stats();
+    }
+
+    @Override
+    public String toString() {
+      return getClass().getSimpleName() + '{' +
+          "expireAfterAccessMs=" + expireAfterAccessMs + ", " +
+          "maxContentLength=" + maxContentLength + ", " +
+          "maxTotalBytes=" + maxTotalBytes + ", " +
+          "cacheStats=" + cache.stats() + '}';
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      CacheEntry buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length : wrappedInputFile.getLength();
+    }
+
+    /**
+     * Opens a new {@link SeekableInputStream} for the underlying data file, either through cache or through the
+     * inner FileIO.
+     * <p>
+     * If data file is not cached yet, and it can fit in the cache, the file content will be cached first before
+     * returning a {@link ByteBufferInputStream}. Otherwise, return a new SeekableInputStream from the inner FIleIO.
+     *
+     * @return a {@link ByteBufferInputStream} if file exist in the cache or can fit in the cache. Otherwise, return a
+     * new SeekableInputStream from the inner FIleIO.
+     */
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        // read-through cache if file length is less than or equal to maximum length allowed to cache.
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        // fallback to non-caching input stream.
+        return wrappedInputFile.newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      }
+    }
+
+    @Override
+    public String location() {
+      return wrappedInputFile.location();
+    }
+
+    @Override
+    public boolean exists() {
+      CacheEntry buf = contentCache.cache.getIfPresent(location());
+      return buf != null || wrappedInputFile.exists();
+    }
+
+    public long getLengthChecked() throws IOException {
+      try {
+        return getLength();
+      } catch (RuntimeIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    private CacheEntry getFileContent() {
+      try {
+        int fileLength = (int) getLength();
```

Comment:
I don't think this should cast to `int`. The length should be tracked as a `long`. I know that we expect `maxContentLength` to be much shorter, but even that is tracked as a `long`. I think this should just use a `long` instead of risking truncation.

Reply:
something bigger than an int long should automatically be excluded from caching, presumably. though if someone did want to declare a cache of > 2GB then it may be needed

Reply:
Changed to `long`.